google/cloud/bigquery/job.py

```diff
@@ -2646,6 +2649,7 @@ def __init__(self, job_id, query, client, job_config=None):
             )
 
         self._query_results = None
+        self._get_query_results_kwargs = {}
```

Comment:
Does this need to be a thread-local variable?

Reply:
Actually, the cached query results might need to be thread-local too. Imagine if two threads called `result` with different starting indexes and/or max results.

Reply:
We'll also need some logic like

https://github.com/googleapis/google-cloud-go/blob/925033712191bce44fa99eb117d6531106042272/bigquery/iterator.go#L314

to see if we can use the cached page if `result` is called more than once

Reply:
Done in latest commit.